frontend/src/layouts/EventsList/EventsList.tsx

```diff
@@ -3,39 +3,121 @@ import { EventsTable } from "../EventsTable/EventsTable";
 import { cn } from "@/lib/utils";
 import styles from "./EventsList.module.css";
 import { TraceEvent } from "@/validators/TraceEvent";
-import { ReactNode } from "react";
+import { ReactNode, useState, useMemo, useEffect, useCallback } from "react";
 import { traceThat } from "tracethat.dev";
+import { useDebounce } from "@/hooks/useDebounce";
+import { EventsSearch } from "@/layouts/EventsSearch/EventsSearch";
+import { usePrevious } from "@/hooks/usePrevious";
+import { useRemount } from "@/hooks/useRemount";
+
+export type EventSearchCriteria = "eventName" | "eventDetails" | "all";
 
 interface EventsListProps {
   data: TraceEvent[];
-  selectedEventCallId: string | null;
   viewerPlaceholder?: ReactNode;
-  onEventClose: () => void;
-  setSelectedEventCallId: (value: string | null) => void;
+  isSearchBarHidden?: boolean;
 }
-export function EventsList({
-  data,
-  selectedEventCallId,
-  setSelectedEventCallId,
-  onEventClose,
-  viewerPlaceholder,
-}: EventsListProps) {
+export function EventsList({ data, viewerPlaceholder, isSearchBarHidden = false }: EventsListProps) {
+  const [selectedEventCallId, setSelectedEventCallId] = useState<string | null>(null);
+
+  const onEventClose = useCallback(() => {
+    setSelectedEventCallId(null);
+  }, [setSelectedEventCallId]);
+
+  const [rawSearchValue, setSearchValue] = useState("");
+  const [searchBy, setSearchBy] = useState<EventSearchCriteria>("all");
+  const searchValue = useDebounce(rawSearchValue, 600);
+
+  const filteredData = useMemo(() => {
+    if (!searchValue) {
+      return data;
+    }
+
+    return data.filter((item) => {
+      const lowerCaseSearchValue = searchValue.toLowerCase();
+
+      const handleDetailsSearch = (details: any) => {
+        return JSON.stringify(details).toLowerCase().includes(lowerCaseSearchValue);
+      };
+
+      if (searchBy === "eventName") {
+        return item.name.toLowerCase().includes(lowerCaseSearchValue);
+      } else if (searchBy === "eventDetails") {
+        return handleDetailsSearch(item.details);
+      } else {
+        return item.name.toLowerCase().includes(lowerCaseSearchValue) || handleDetailsSearch(item.details);
+      }
+    });
+  }, [searchValue, searchBy, data]);
+
+  const searchValueByName = searchBy !== "eventDetails" ? searchValue : "";
+  const searchValueByDetails = searchBy !== "eventName" ? searchValue : "";
+
+  const isSearchNotFound = searchValue && filteredData.length === 0;
+
+  const previousFirst = usePrevious(filteredData[0]);
+  const previousSearchValue = usePrevious(searchValue);
+
+  useEffect(() => {
+    if (!filteredData[0]) return;
+
+    if (filteredData[0] != previousFirst || previousSearchValue != searchValue) {
+      setSelectedEventCallId(filteredData[0].callId);
+    }
+  }, [filteredData[0], previousFirst, searchValue, previousSearchValue]);
+
+  const { mounted: viewerMounted, remount: remountViewer } = useRemount();
```

Comment:
As it's not a standard I would put a comment about why it's needed e.g. "We want the event viewer should reset it's internal state when the user is searching, to highlight the matching phrases."